### PR TITLE
React 18 final support

### LINKED ISSUE #142: Getting warning from ReactDOM when using React 18
Hello,

When upgrading to React 18 and using single-spa-react with the following configuration:

```javascript
import React from 'react';
import ReactDOM from 'react-dom';
import singleSpaReact from 'single-spa-react';

import Root from './root.component';

const lifecycles = singleSpaReact({
  React,
  ReactDOM,
  rootComponent: Root,
  renderType: 'createRoot',
});
```

I get the following warning in the console: `Warning: You are importing createRoot from "react-dom" which is not supported. You should instead import it from "react-dom/client".`

However, the single-spa-react api is requiring ReactDOM to be passed. Would it be better to pass the `createRoot` function here or drop the requirement of passing ReactDOM when using React 18?

### PR DESCRIPTION
**Pull request by filoxo**
Resolves #142 and #115. The APIs for React 18 had been implemented as communicated at the time, but the final release has some other changes made to it that require updating. Namely, `hydrateRoot` behavior and the removal of `createBlockingRoot`. This PR is a **breaking change** unfortunately because I couldn't find a way to conditionally apply types based on package version; as well as defaulting to `createRoot`. This PR also includes updated tests to use React's `act` utility and some other minor corrections to package metadata. I've tested this with two React applications generated with create-single-spa, one with TypeScript and one without. Below are the changes needed once the new version is installed.

```diff
import React from "react";
-import ReactDOM from "react-dom";
+import ReactDOMClient from "react-dom/client";
import singleSpaReact from "single-spa-react";
import Root from "./root.component";

const lifecycles = singleSpaReact({
  React,
- ReactDOM,
+ ReactDOMClient,
  rootComponent: Root,
- renderType: "render" // this should only be necessary if renderType had been explicitly defined
+ renderType: "createRoot"
});
```

### CODE REVIEW

File: types/single-spa-react.d.ts [hunk 1]
```diff
@@ -1,21 +1,23 @@
...
-    | React.FunctionComponent<RootComponentProps>;
-  loadRootComponent?: (
-    props: RootComponentProps
-  ) => Promise<
-    | React.ComponentClass<RootComponentProps, any>
-    | React.FunctionComponent<RootComponentProps>
-  >;
+  ReactDOMClient: typeof ReactDOMClient;
+  rootComponent?: React.ElementType<RootComponentProps>;
+  loadRootComponent?: (props?: any) => Promise<React.ElementType<any>>;
```
Comment (filoxo) [inline types/single-spa-react.d.ts:18]: I encountered a strange TypeScript-ism with the previous typings for this. Because they are using the same type, if the outer function doesn't have props, the type is inferred as `never` which then gets collapsed/passed/inferred down (idk what to call this behavior) to the props being passed to the component even though you might not need them. More concretely, it means you cannot have inner props defined on the function that the Promise resolves with if the outer props aren't also defined.

```js
loadRootComponent: () => Promise.resolve(() => <div>hi</div>)) // OK
loadRootComponent: (outerProps) => Promise.resolve((innerProps) => <div>hi</div>)) // OK
loadRootComponent: (outerProps) => Promise.resolve(() => <div>hi</div>)) // OK
loadRootComponent: () => Promise.resolve((innerProps) => <div>hi</div>)) // ERROR
```

Loosening the type here meant that you can now use that last combination without error. This wasn't worth the time I put into it but its done now.

File: src/single-spa-react.js [hunk 2]
```diff
@@ -207,30 +191,19 @@ function atLeastReact16(React) {
...
-      "unstable_createRoot",
-      "createBlockingRoot",
-      "unstable_createBlockingRoot",
-    ].indexOf(renderType) >= 0
-  ) {
-    const root = opts.ReactDOM[renderType](domElement);
-    root.render(elementToRender);
+
+  if (renderType === "hydrateRoot") {
+    const root = opts.ReactDOMClient[renderType](domElement, elementToRender);
```
Comment (filoxo) [inline src/single-spa-react.js:196]: The migration docs made this specific callout:

> // Unlike with createRoot, you don't need a separate root.render() call here.

See also: https://reactjs.org/blog/2022/03/08/react-18-upgrade-guide.html#updates-to-client-rendering-apis

File: src/single-spa-react.js [hunk 3]
```diff
@@ -207,30 +191,19 @@ function atLeastReact16(React) {
...
-    // default to this if 'renderType' is null or doesn't match the other options
-    opts.ReactDOM.render(elementToRender, domElement);
+  switch (renderType) {
+    case "createRoot": {
+      const root = opts.ReactDOMClient.createRoot(domElement);
+      root.render(elementToRender);
+      return root;
+    }
+    case "hydrateRoot": {
+      const root = opts.ReactDOMClient.hydrateRoot(domElement, elementToRender);
```
Comment (filoxo) [inline src/single-spa-react.js:201]: The migration docs made this specific callout:

> // Unlike with createRoot, you don't need a separate root.render() call here.

See also: https://reactjs.org/blog/2022/03/08/react-18-upgrade-guide.html#updates-to-client-rendering-apis

**Review by TheMcMurder — APPROVED** [review]
Looks good to me. Thanks for all the comments to help me understand the changes.

**Comment by filoxo** [comment]
Quick update. I still have to write the updated documentation for single-spa-react. This I'll be out of town for this week so I may try to publish an alpha, but no guarantees.

**Comment by Alykod** [comment]
Any updates on when this PR will be merged? :) Ty

**Comment by felipe-soares-fortis** [comment]
Any updates on when this update will be released?

**Comment by seanmcquaid** [comment]
Hate to pile on here but also curious when this might get released??

**Review by KhanShaheb34 — APPROVED** [review]
Please merge these changes ASAP.

**Review by joeldenning — COMMENTED** [review]
In previous versions, I've tried quite hard to have one version of single-spa-react compatible with all versions of React. I believe that single-spa-react@4 works with react 0.15 through react 17. The reason is that it is a pretty big pain to backport any bug fixes to old versions of single-spa-react, and that backporting usually just doesn't happen. The single-spa-angular project has different major versions for each version of Angular, and it is a pretty big pain.

I'd prefer continuing this for React 18, if it's possible and doesn't require a ton of work. What are your thoughts on that? The `ReactDOMClient` change to the single-spa-react api is the first one that stuck out to me as something probably intended only for React 18 and not for React 17. But reading through the code, it looks like the main thing is just handling `hydrateRoot` differently than `createRoot`, and that seems like something we can do without a breaking change to single-spa-react nor creating separate release lines for React 18 versus React 15-17

**Comment by joeldenning** [comment]
> Any updates on when this PR will be merged? :) Ty
> Any updates on when this update will be released?
> Please release these changes ASAP. 🥹

FYI comments like this generally make me not want to work on something. I know they're not meant with bad intentions, but after five years of my expert labor being leeched for free by thousands of companies, it is hard for me to not to be spiteful towards the people trying to get me to do things.

File: types/single-spa-react.d.ts [hunk 4]
```diff
@@ -1,21 +1,21 @@
 import * as React from "react";
-import * as ReactDOM from "react-dom";
+import * as ReactDOMClient from "react-dom/client";
 import { AppProps, CustomProps, LifeCycleFn } from "single-spa";
 
 export const SingleSpaContext: React.Context<CustomProps & AppProps>;
 
+export interface ReactAppOrParcel<ExtraProps> {
```
Comment (joeldenning) [inline types/single-spa-react.d.ts:7]: Why is this needed? Could we use single-spa's `Lifecycles` type instead?

https://github.com/single-spa/single-spa/blob/80682e10ec929644d7a50280ea0d43675d31ee56/typings/single-spa.d.ts#L56

File: types/single-spa-react.d.ts [hunk 5]
```diff
@@ -1,21 +1,21 @@
...
-  rootComponent?:
-    | React.ComponentClass<RootComponentProps, any>
-    | React.FunctionComponent<RootComponentProps>;
-  loadRootComponent?: (
-    props: RootComponentProps
-  ) => Promise<
-    | React.ComponentClass<RootComponentProps, any>
-    | React.FunctionComponent<RootComponentProps>
-  >;
+  ReactDOMClient: typeof ReactDOMClient;
```
Comment (joeldenning) [inline types/single-spa-react.d.ts:16]: As far as I know, typescript doesn't have a way for us to provide the correct types for these two properties. The single-spa-react project supports React 15-18, which has a variety of types. In an ideal world, we'd do something like this:

```ts
import * as React15 from 'react15';
import * as React16 from 'react16';
import * as React17 from 'react17';
import * as React18 from 'react18';
import * as ReactDOM15 from 'react15';
import * as ReactDOM16 from 'react16';
import * as ReactDOM17 from 'react17';
import * as ReactDOM18 from 'react18';

export interface SingleSpaReactOpts {
  React: typeof React15 | typeof React16 | typeof React17 | typeof React18;
  ReactDOM: typeof ReactDOM15 | typeof ReactDOM16 | typeof ReactDOM17 | typeof ReactDOM18;
}
```

However, I don't think it's possible to install all the different versions of `@types/react` and then get typescript to properly alias them in a generally reusable way.

For this reason, i propose just switching to `any` for the `React` and `ReactDOM` types

**Comment by filoxo** [comment]
Thanks for the feedback Joel! I'm reworking this to avoid the breaking change.

File: types/single-spa-react.d.ts (hunk 5 above)
Comment (limulus) [reply types/single-spa-react.d.ts:16]: One alternative to using `any` might be to use like types for `ReactDOM`. Maybe something along these lines?

```typescript
interface ReactDOMLike {
  render: typeof ReactDOM.render;
}

interface ReactDOMClientLike {
  createRoot: typeof ReactDOMClient.createRoot;
  hydrate: typeof ReactDOMClient.hydrateRoot;
}

export interface SingleSpaReactOpts {
  ReactDOM: ReactDOMLike | ReactDOMClientLike;
}
```

Comment (filoxo) [reply types/single-spa-react.d.ts:16]: @limulus how would that work if I'm using React 17 still and my version of react types doesn't have `ReactDOMClient`? The whole point is that we want backwards compatibility so using ReactDOMClient type is not possible without making single-spa-react incompatible with react 17 and below. At least that's my understanding.

Comment (limulus) [reply types/single-spa-react.d.ts:16]: @filoxo Yeah, you are right, as I wrote it that would not work. I was going to suggest that you could inline the types of the `createRoot`, `hydrate`, and `render` functions so that you would not need to pull those types from the React types package. But I took a look at your solution from the other PR and that seems reasonable enough to me!
